### PR TITLE
Fix unbounded memory growth for Ruby <2.7.

### PR DESCRIPTION
**Pull request by haberman**
Previously our secondary map (needed only on Ruby <2.7) was growing without bound.

This PR adds a periodic GC pass of the secondary map to avoid this unbounded memory growth.

Fixes: https://github.com/protocolbuffers/protobuf/issues/8421

### CODE REVIEW

File: ruby/ext/google/protobuf_c/protobuf.c [hunk 1]
```diff
@@ -251,16 +251,65 @@ void Arena_register(VALUE module) {
...
+ID length;
+
+extern VALUE weak_obj_cache;
+
 static void SecondaryMap_Init() {
   rb_gc_register_address(&secondary_map);
+  rb_gc_register_address(&gc_secondary_map);
+  rb_gc_register_address(&secondary_map_mutex);
   secondary_map = rb_hash_new();
+  gc_secondary_map = rb_eval_string(
```
Comment (fowles) [inline ruby/ext/google/protobuf_c/protobuf.c:272]: I didn't realize this was a functor that did the gc, I thought it was a map.  Can we name it something more evocative:
gc_secondary_map_functor or something?

File: ruby/ext/google/protobuf_c/protobuf.c [hunk 2]
```diff
@@ -251,16 +251,65 @@ void Arena_register(VALUE module) {
...
+// The main object cache is a WeakMap that will automatically remove entries
+// when the target object is no longer reachable, but unless we manually
+// remove the corresponding entries from the secondary map, it will grow
+// without bound.
+//
+// To avoid this unbounded growth we periodically remove entries from the
+// secondary map that are no longer present in the WeakMap. The logic of
+// how often to perform this GC is an artbirary tuning parameter that
+// represents a straightforward CPU/memory tradeoff.
+static void SecondaryMap_MaybeGC() {
```
Comment (fowles) [inline ruby/ext/google/protobuf_c/protobuf.c:290]: I believe that one must hold the mutex to call this function.  Can you add that as a comment and some kind of assertion as well?

Comment (haberman) [reply ruby/ext/google/protobuf_c/protobuf.c:292]: Done.

File: ruby/ext/google/protobuf_c/protobuf.c (hunk 1 above)
Comment (haberman) [reply ruby/ext/google/protobuf_c/protobuf.c]: Done.

**Review by fowles — APPROVED** [review]

**Pull request merged**